### PR TITLE
add timestamp to ethr builder

### PR DESCRIPTION
**Pull request by insipx**
related to https://github.com/xmtp/xps-gateway/issues/61

timestamp is not the best way b/c we can have multiple installation id updates in the same block. It may be OK in the general case, since it could be rare that a user updates their installation key in the same block twice, but it can happen. libxmtp relies on it in order to sort through installation ids it already has. In the future it would be better to use a block # / log index combo. This would also avoid yet another api call to the ethereum rpc node

### CODE REVIEW

**Review by 37ng — APPROVED** [review]
LGTM!
Looks like we need a rustfmt.toml to unify `cargo fmt`

**Comment by insipx** [comment]
> LGTM! Looks like we need a rustfmt.toml to unify `cargo fmt`

we should be just on the stable default rustfmt settings, in this case my lsp crashed and i didn't get the autoformatting I usually do

File: lib/src/resolver.rs [hunk 1]
```diff
@@ -28,6 +28,33 @@ impl<M> From<DIDRegistry<M>> for Resolver<M> {
     }
 }
 
+/// Extra context passed to the document builder from the [`Resolver`]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EventContext {
+    /// the timestamp in nanoseconds in which the block from the document was built.
+    pub timestamp: i64,
```
Comment (jac18281828) [inline lib/src/resolver.rs:35]: Block timestamp may be more relevant for future viewers of this code

File: lib/src/resolver.rs [hunk 2]
```diff
@@ -28,6 +28,33 @@ impl<M> From<DIDRegistry<M>> for Resolver<M> {
...
+impl EventContext {
+    pub async fn new<M: Middleware>(
+        meta: &LogMeta,
+        signer: impl Middleware,
+    ) -> Result<Self, ResolverError<M>> {
+        let block = signer
+            .get_block(meta.block_number)
+            .await
+            .map_err(|e| ResolverError::Middleware(e.to_string()))?;
+        let timestamp = block
```
Comment (jac18281828) [inline lib/src/resolver.rs:47]: This should be consistent by fetching the block time from Ethereum or failing.   Block time can differ from real time.

**Review by jac18281828 — CHANGES_REQUESTED** [review]
I think we should be consistent in our use of block timestamp in seconds

**Comment by insipx** [comment]
removed logic to convert to datetime b/c it was wasteful, changed to store the usual timestamp and then multiply for nanoseconds @jac18281828

File: lib/src/types/xmtp.rs [hunk 3]
```diff
@@ -83,12 +86,18 @@ impl EthrBuilder {
         index: usize,
         value: V,
         key: XmtpAttribute,
+        context: &EventContext,
     ) -> Result<(), EthrBuilderError> {
         log::debug!("index: {}", index);
+
+        let timestamp_ns = context.block_timestamp * NANOSECONDS_PER_SECOND;
```
Comment (jac18281828) [inline lib/src/types/xmtp.rs:93]: https://doc.rust-lang.org/std/time/struct.Duration.html#method.as_nanos

Comment (jac18281828) [reply lib/src/types/xmtp.rs:93]: Is use Duration for this

**Review by jac18281828 — APPROVED** [review]
Looks good !  One last thing, I’d use Duration for time conversion.

**Pull request merged**